refactor!: enshrine note randomness

mverzilli opened the pull request:
This PR makes Aztec-nr handle note randomness as a built-in feature.

Before this PR, randomness injection was a userland concern: users of Aztec-nr were expected to remember to include a randomness attribute in the body of their custom notes to protect against pre-image attacks.

After this PR, the randomness is part of the note metadata Aztec-nr handles natively (similar to what happens with storage slot, for example).

Fixes F-32

mverzilli commented on noir-projects/aztec-nr/aztec/src/messages/discovery/partial_notes.nr:
```diff
@@ -182,16 +187,17 @@ fn decode_partial_note_private_msg(
     let storage_slot = msg_content.get(0);
-    let note_completion_log_tag = msg_content.get(1);
+    let randomness = msg_content.get(1);
+    let note_completion_log_tag = msg_content.get(2);
```
Would be nice not to have the knowledge of "you find this enshrined field at this index" spread around random places like this

mverzilli replied:
for now I put the indexes in constants, but still it would be nice to have a single place where we spec these wire formats out and use tooling to generate implementations

mverzilli commented on noir-projects/aztec-nr/aztec/src/messages/discovery/partial_notes.nr:
```diff
@@ -182,16 +187,17 @@ fn decode_partial_note_private_msg(
+    let note_completion_log_tag = msg_content.get(2);
 
-    let packed_private_note_content = array::subbvec(msg_content, 2);
+    let packed_private_note_content = array::subbvec(msg_content, 3);
```
ditto

mverzilli commented on noir-projects/aztec-nr/aztec/src/messages/discovery/private_notes.nr:
```diff
@@ -101,13 +105,14 @@ fn decode_private_note_msg(
 
     // We currently have a single field that is not the note's packed representation, which is the storage slot.
     let storage_slot = msg_content.get(0);
+    let randomness = msg_content.get(1);
```
TODO: do sth nicer about this hardcodes

mverzilli commented on noir-projects/aztec-nr/aztec/src/messages/logs/note.nr:
```diff
@@ -32,17 +34,19 @@ where
     storage_slot: Field,
-) -> [Field; <Note as Packable>::N + 2]
+    randomness: Field,
+) -> [Field; <Note as Packable>::N + 3]
```
kill this magic constant (need to check if Noir's type grammar let's me use a constant here though)

mverzilli commented on noir-projects/aztec-nr/aztec/src/messages/logs/note.nr:
```diff
@@ -32,17 +34,19 @@ where
+    msg_content[1] = randomness;
     for i in 0..packed_note.len() {
-        msg_content[1 + i] = packed_note[i];
+        msg_content[2 + i] = packed_note[i];
```
ditto

mverzilli commented on noir-projects/aztec-nr/aztec/src/messages/logs/note.nr:
```diff
@@ -52,17 +56,19 @@ where
     storage_slot: Field,
-) -> [Field; <PartialNotePrivateContent as Packable>::N + 2]
+    randomness: Field,
+) -> [Field; <PartialNotePrivateContent as Packable>::N + 3]
```
ditto

mverzilli commented on noir-projects/aztec-nr/aztec/src/messages/logs/note.nr:
```diff
@@ -52,17 +56,19 @@ where
+    msg_content[1] = randomness;
     for i in 0..packed_private_content.len() {
-        msg_content[1 + i] = packed_private_content[i];
+        msg_content[2 + i] = packed_private_content[i];
```
ditto

mverzilli commented on yarn-project/pxe/src/contract_function_simulator/noir-structs/note_validation_request.ts:
```diff
@@ -4,7 +4,7 @@ import { AztecAddress } from '@aztec/stdlib/aztec-address';
 
 // TODO(#14617): should we compute this from constants? This value is aztec-nr specific.
-export const MAX_NOTE_PACKED_LEN = 12;
+export const MAX_NOTE_PACKED_LEN = 11;
```
spent hours to hunt down e2e failing tests due to not updating this value. need to think of a some way of making it fail more spectacularly and faster

mverzilli commented on yarn-project/pxe/src/contract_function_simulator/oracle/note_packing_utils.ts:
```diff
@@ -48,5 +51,5 @@ export function packAsRetrievedNote({
 
   // Pack metadata first (stage and maybe_note_nonce), followed by the rest
-  return [...note.items, contractAddress, new Fr(noteMetadata.stage), noteMetadata.maybeNoteNonce];
+  return [...note.items, contractAddress, randomness, new Fr(noteMetadata.stage), noteMetadata.maybeNoteNonce];
```
this is more subtle but it's also kind of "magic constant-ish" there's a bespoke way of packing a retrieved note

mverzilli commented on yarn-project/pxe/src/private_kernel/private_kernel_execution_prover.test.ts:
```diff
@@ -50,7 +50,15 @@ describe('Private Kernel Sequencer', () => {
+          new Note([Fr.random(), Fr.random(), Fr.random()]),
+          Fr.random(),
+          Fr.random(),
+          NoteSelector.random(),
```
these kinds of lengthy parameter lists where multiple parameters are of the same type are always a problem waiting to occur in JS, consider replacing with object with named attrributes

mverzilli commented on yarn-project/stdlib/src/tx/private_execution_result.ts:
```diff
@@ -35,17 +38,18 @@ export class NoteAndSlot {
 
   static random() {
-    return new NoteAndSlot(Note.random(), Fr.random(), NoteSelector.random());
+    return new NoteAndSlot(Note.random(), Fr.random(), Fr.random(), NoteSelector.random());
```
ditto

mverzilli replied:
having second thoughts about this change. leaving aside that I should rename the type (it's no longer _just_ a note and slot), I'm not 100% sure that we should show randomness here

benesjan commented on noir-projects/aztec-nr/aztec/src/messages/discovery/partial_notes.nr:
```diff
@@ -47,18 +52,9 @@ pub unconstrained fn process_partial_note_private_msg(
-        packed_private_note_content,
-        recipient,
-    };
+    let pending = decode_partial_note_private_msg(msg_metadata, msg_content, recipient);
```
It took me a bit of time to figure out the changes here so if you could put here a github comment next time that would be helpful. Not a big deal though. Returning the `DeliveredPendingPartialNote` directly from `decode_partial_note_private_msg` function makes sense 👍

benesjan commented on noir-projects/aztec-nr/aztec/src/messages/logs/note.nr:
```diff
@@ -11,10 +11,15 @@ use protocol_types::{
 
+global PRIVATE_NOTE_RESERVED_FIELDS: u32 = 2;
+global PRIVATE_NOTE_STORAGE_SLOT_INDEX: u32 = 0;
+global PRIVATE_NOTE_RANDOMNESS_INDEX: u32 = 1;
```
Good call on introducing these constants 👍

benesjan commented on noir-projects/aztec-nr/aztec/src/messages/logs/note.nr:
```diff
@@ -32,17 +38,20 @@ where
     storage_slot: Field,
-) -> [Field; <Note as Packable>::N + 2]
+    randomness: Field,
+) -> [Field; <Note as Packable>::N + PRIVATE_NOTE_RESERVED_FIELDS + 1]
```
Given that you've already introduced the constants here I think it would make sense to replace the "+1" with the `MESSAGE_EXPANDED_METADATA_LEN` constant that can be exported from the `encoding.nr` file.

benesjan commented on noir-projects/aztec-nr/aztec/src/messages/logs/note.nr:
```diff
@@ -52,17 +61,21 @@ where
     storage_slot: Field,
-) -> [Field; <PartialNotePrivateContent as Packable>::N + 2]
+    randomness: Field,
+) -> [Field; <PartialNotePrivateContent as Packable>::N + PRIVATE_NOTE_RESERVED_FIELDS + 1]
```
Same here

benesjan commented on noir-projects/aztec-nr/aztec/src/test/helpers/test_environment.nr:
```diff
@@ -754,17 +772,18 @@ impl TestEnvironment {
         // typically be created by the macros for the different notes in a given contract. Here we build one specialized
         // for `Note`.
-        let compute_note_hash_and_nullifier: ComputeNoteHashAndNullifier<_> = |packed_note, storage_slot, note_type_id, contract_address, note_nonce| {
+        let compute_note_hash_and_nullifier: ComputeNoteHashAndNullifier<_> = |packed_note, storage_slot, note_type_id, contract_address, randomness: Field, note_nonce| {
```
```suggestion
        let compute_note_hash_and_nullifier: ComputeNoteHashAndNullifier<_> = |packed_note, storage_slot, note_type_id, contract_address, randomness, note_nonce| {
```

Style-wise it just looked strange to have the type explicitly set only for randomness.

mverzilli replied:
Nice catch. This one's funny: I get confused with the LSP which adds the inferred types so sometimes I end up typing them :sweat_smile:

benesjan commented on noir-projects/noir-contracts/contracts/app/token_blacklist_contract/src/main.nr:
```diff
@@ -48,6 +48,9 @@ pub contract TokenBlacklist {
     // Changing an address' roles has a certain delay before it goes into effect. Here set to 1 day.
     global CHANGE_ROLES_DELAY: u64 = 86400;
 
+    // We don't want randomness for transparent notes, but we need to provide a value.
```
```suggestion
    // Randomness is unnecessary for transparent notes, as these notes are emitted publicly during the shielding of
    // tokens, making protection against preimage attacks irrelevant. However, the API requires a value, so we use 0.
```

Think it's helpful to provide more context here.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/noir-structs/note_validation_request.ts:
```diff
@@ -38,9 +40,16 @@ export class NoteValidationRequest {
     const txHash = TxHash.fromField(reader.readField());
     const recipient = AztecAddress.fromField(reader.readField());
 
+    if (reader.remainingFields() !== 0) {
```
very nice 👍

benesjan reviewed (approved):
Very nice! Great job on eating the glass and getting it all to pass.

Before merging this, please write migration notes. Devs can define their own notes or they can be using our note implementations (e.g. `UintNote`) so this affects them. The migration notes are written in `docs/docs/developers/docs/resources/migration_notes.md` and you should put the changes into the "TBD" section - specific version there gets set when a new release is made.

Since the changes in partial notes are quite complex I think you could get by just by linking to this PR and telling devs to look at the diff of some implementation of a partial note.

Also marked the PR as breaking change by putting "!" after "refactor" in the title.

mverzilli commented:
> Before merging this, please write migration notes.

Thanks for explaining this! I just added the corresponding migration notes. Considering it's a breaking change, I wonder if I should change the title from `refactor!` to `feat!`. From the point of view of Aztec.nr this is a new feature that users used to have to implement themselves.

mverzilli committed:
refactor!: enshrine note randomness

This PR makes Aztec-nr handle note randomness as a built-in feature.

Before this PR, randomness injection was a userland concern: users of Aztec-nr were expected to remember to include a randomness attribute in the body of their custom notes to protect against pre-image attacks.

After this PR, the randomness is part of the note metadata Aztec-nr handles natively (similar to what happens with storage slot, for example).

Fixes F-32

Co-authored-by: benesjan <janbenes1234@gmail.com>
